src/utils/SocketClient.hxx

```diff
@@ -0,0 +1,243 @@
+ * @date 11 March 2017
+ */
+
+#ifndef _UTILS_SOCKET_CLIENT_HXX_
+#define _UTILS_SOCKET_CLIENT_HXX_
+
+#include <functional>
+#include <netdb.h>
+#include <arpa/inet.h>
+
+#include "executor/StateFlow.hxx"
+#include "executor/Timer.hxx"
+
+class SocketClient : public StateFlowBase, public OSThread , public OSSem
+{
+public:
+    /** Constructor.
+     * @param service service that the StateFlowBase will be bound to.
+     * @param mdns_name service name to connect to, nullptr to force use
+     *                  hostsname and port
+     * @param host host to connect to if mdns_name resolution fails,
+     *             nullptr to force use mDNS
+     * @param port port number to connect to if mdns_name resolution fails
+     * @param callback callback method to invoke when a client connection is
+     *                 made successfully.  It is the responsibility of the
+     *                 callee to register the notifiable (on close) if this
+     *                 client shall reattempt the connection if the socket
+     *                 is ever closed.
+     * @param retry_seconds time in seconds that the client shall wait to retry
+     *                      connecting on error.
+     */
+    SocketClient(Service *service, const char *mdns, const char * host,
+                 int port,
+                 std::function<void(int, char*, int, Notifiable*)>callback,
+                 unsigned retry_seconds = 5)
+        : StateFlowBase(service)
+        , OSThread()
+        , OSSem()
+        , mdns_(mdns)
+        , host_(host)
+        , port_(port)
+        , callback_(callback)
+        , retrySeconds_(retry_seconds)
+        , selectHelper_(this)
+        , timer_(this)
+        , fd_(-1)
+        , addr_(nullptr)
+    {
+        HASSERT(mdns_ || (host_ && port_));
+        start_flow(STATE(spawn_thread));
+    }
+
+private:
+    /** thread that will handle the blocking address resolution.
+     * @return should never return
+     */
+    virtual void *entry() override
+    {
+        int ai_ret;
+        struct addrinfo hints;
+
+        memset(&hints, 0, sizeof(hints));
+        hints.ai_family = AF_INET;
+        hints.ai_socktype = SOCK_STREAM;
+        hints.ai_flags = 0;
+        hints.ai_protocol = IPPROTO_TCP;
+
+        for ( ; /* forever */ ; )
+        {
+            OSSem::wait();
+            for ( ; /* forever */ ; )
+            {
+                if (mdns_)
+                {
+                    /* mdns address resolution */
+                    ai_ret = getaddrinfo(nullptr, mdns_, &hints, &addr_);
+                }
+                else
+                {
+                    char port_str[30];
+                    integer_to_buffer(port_, port_str);
+                    ai_ret = getaddrinfo(host_, port_str, &hints, &addr_);
+                }
+                if (ai_ret == 0 && addr_)
+                {
+                    notify();
+                    break;
+                }
+                if (addr_)
```

Comment:
make addr_ an std::unique_ptr<struct addr_info, freeaddrinfo>   (makes sure to use the deleter of freeaddrinfo).